server/src/modules/maps/maps.service.ts

```diff
@@ -440,6 +449,226 @@ export class MapsService {
+                trackNum: 0,
+                zoneNum: 0
+            }
+        };
+
+        if (query.flags) where.flags = query.flags;
+        if (query.trackNum) where.run.trackNum = query.trackNum;
+        if (query.zoneNum) where.run.zoneNum = query.zoneNum;
+
+        const include = {
+            run: true
+        };
+
+        const order: Prisma.UserMapRankOrderByWithAggregationInput = {
+            rank: 'asc'
+        };
+
+        const myRankInfo = await this.mapRepo.getRank(where, include);
+
+        if (!myRankInfo) throw new NotFoundException('No personal best detected');
+
+        const myRank = myRankInfo.rank;
+
+        where.userID = undefined;
+
+        const ranks = await this.mapRepo.getRanks(where, include, order, Math.max(myRank - 6, 0), 11);
+
+        //This is done because the MapRankDto still contains trackNum and zoneNum but UserMapRank doesn't
+        const ret = ranks[0].map((item) =>
+            DtoFactory(MapRankDto, {
+                mapID: item.mapID,
+                userID: item.userID,
+                runID: item.runID,
+                gameType: item.gameType,
+                flags: item.flags,
+                trackNum: (item as any).run.trackNum,
+                zoneNum: (item as any).run.zoneNum,
+                rank: item.rank,
+                rankXP: item.rankXP,
+                createdAt: item.createdAt,
+                updatedAt: item.updatedAt
+            })
+        );
+
+        return ret;
+    }
+
+    async getRankFriends(userID: number, mapID: number, query: MapRankGetNumberQuery): Promise<MapRankDto[]> {
+        const user = await this.userRepo.get(userID);
+        const steamID = user.steamID;
+
+        const res = await firstValueFrom(
+            this.http
+                .get('https://api.steampowered.com/ISteamUser/GetFriendList/v1/', {
+                    params: {
+                        key: this.config.get('steam.webAPIKey'),
+                        steamID: steamID,
+                        relationship: 'friend'
+                    }
+                })
+                .pipe(
+                    catchError((error: AxiosError) => {
+                        if (error.response.status === 401)
+                            throw new ConflictException('Friends list or profile is private');
+
+                        throw new BadRequestException('Bad request');
+                    })
+                )
+        );
+
+        if (res.data) {
+            const friendSteamIDs = res.data.friendslist.friends.map((item) => item.steamid);
+
+            if (friendSteamIDs.length === 0) throw new ImATeapotException('No friends detected :(');
+
+            //This must be done like this because mapping promises is wierd
+            const friendPromises = friendSteamIDs.map((item) => this.userRepo.getBySteamID(item));
+            const friends: User[] = await Promise.all(friendPromises);
+            const friendIDs = friends.map((item) => item.id);
```

Comment:
Funny comment here, this is a very normal use of promises in JS. This is the standard way of parallelising a bunch of stuff, i.e. make an array of promises and await them all - `Promise.all` returns a new promise that runs an array of promises in parallel and resolves when they've all resolved.

That being said, this isn't the right way to do this. Rather than make a bunch of separate DB queries ourselves, it's best to just let Prisma handle it for us, as there's likely optimisations made that'll make it run faster.

So instead, we can do something like `const friends = await this.userRepo.getAll({ steamID: { in: friendSteamIDs } }, {});` (for some reason getAll requires an `include` object, IMO would be worth changing that, just making the include optional. but commit separately if you do that)

Reply:
Actually, why are you doing this at all? SteamIDs are unique to `User`s in the database, just use that in your `where`. So just scrap all this and change your getRanks `where` to use `steamID`s instead of `userID`s.